### PR TITLE
[FIREBASE] - Added firebase support and reshaped fragments

### PR DESCRIPTION
**Pull request by cosmin-ionita**
@radumarin88: This PR has the following changes:

- I finished the Home fragment which contains general information, like: upcoming exams / assignments and a graph with attendance distribution over few months (currently this fragment has static data)
- I finished all the other views and I linked them with Firestore
- I also integrated email & password authentication with FirebaseAuth
- I also integrated FirestoreRecyclerAdapter from firestore-ui to be able to fetch the data easier
- I added a new fragment that writes to Firestore (the only one actually). It's a simple form for document requests.

### CODE REVIEW

File: app/src/main/java/android/example/courses/AttendanceFragment.kt [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+    private var attendanceAdapter: FirestoreRecyclerAdapter<Attendance, AttendanceViewHolder>? =
+        null
+
+    override fun onCreateView(
+        inflater: LayoutInflater, container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
+
+        val query = db.collection("attendance/${auth.currentUser?.email}/data")
```
Comment (cosmin-ionita) [inline app/src/main/java/android/example/courses/AttendanceFragment.kt]: @radumarin88: When I tried to read data from firestore and to display it with a recycler view I was a bit confused by this query thing.

I kind of makes sense the way it works: you pass in a query that should return multiple documents and for each document the method  `onBindViewHolder` is called.

But what if you have a document that has an array or a map (as a field) and you want to display that array/map in the recycler view? In that case, you'll basically return one document each time, so you won't be able to build the query (or at least I don't know how to do it).

I'm asking this because initially my data model was the following:

`attendance (collection) --> user@email.com (document) --> attendace: [course1: 4, course2: 7, course3: 10]`

So I just wanted to say `.get(user@email.com)`, to get the `attendance` field and to use it to power up the recyclerView.

I didn't work, so I had to update the data model to:

`attendance (collection) --> user@email.com (document) --> data (collection) --> courseName (document) -> fields `

Basically, I had to add an extra collection (named `data` in that case) and inside it I added a document for each course.

What do you think about that? Is this the right way of doing it?

File: app/src/main/java/android/example/courses/CourseDetailFragment.kt [hunk 2]
```diff
@@ -5,16 +5,59 @@ import androidx.fragment.app.Fragment
...
-        /* Load the details of the course given by that ID */
-
+        selectedCourse = arguments?.get("courseName") as String
         return inflater.inflate(R.layout.fragment_course_detail, container, false)
     }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+//        val loadingScreen = view.findViewById<FrameLayout>(R.id.progress_overlay)
```
Comment (cosmin-ionita) [inline app/src/main/java/android/example/courses/CourseDetailFragment.kt:28]: @radumarin88 Here I tried to add a loading screen that would show up when the data is loading (It didn't seem to work - the screen remained blank after the data was loaded)

File: app/src/main/java/android/example/courses/NavigationActivity.kt [hunk 3]
```diff
@@ -31,14 +39,51 @@ class NavigationActivity : AppCompatActivity(),
...
+        var email = auth.currentUser?.email!!
+
+        db.collection("users").document(email)
+            .get()
+            .addOnSuccessListener {document ->
+                if (document != null) {
+                    nameView.text = (document.data?.get("name") as String)
+
+                    val link = (document.data?.get("pictureUrl") as String)
+                    Picasso.get()
```
Comment (cosmin-ionita) [inline app/src/main/java/android/example/courses/NavigationActivity.kt:77]: I saw that on a slow network the profile picture is downloaded in a few seconds, during which is not displayed at all. How can I fix that? Should I put a loading screen right after the login step? If so, how can I know that all the data was loaded completely to be able to hide it?

File: app/build.gradle [hunk 4]
```diff
@@ -46,4 +46,16 @@ dependencies {
...
+    implementation 'com.google.firebase:firebase-firestore:21.3.0'
+    implementation 'com.squareup.picasso:picasso:2.71828'
+    implementation "org.jetbrains.kotlinx:kotlinx-coroutines-core:1.3.3"
+    implementation "org.jetbrains.kotlinx:kotlinx-coroutines-android:1.3.3"
+    implementation 'com.firebaseui:firebase-ui-firestore:6.2.0'
+    implementation 'com.firebaseui:firebase-ui-auth:6.2.0'
+    implementation 'com.pnikosis:materialish-progress:1.7'
 }
+
+apply plugin: 'com.google.gms.google-services'
```
Comment (radumarin88) [inline app/build.gradle:61]: You didn't add the google-services.json in the repo (I know it should be secure), so I can't build the project. Or you could add me to your firebase console so I can download it myself.

File: app/proguard-rules.pro [hunk 5]
```diff
@@ -6,7 +6,7 @@
 #   http://developer.android.com/guide/developing/tools/proguard.html
 
 # If your project uses WebView with JS, uncomment the following
-# and specify the fully qualified class name to the JavaScript interface
+# and specify the fully qualified class courseName to the JavaScript interface
```
Comment (radumarin88) [inline app/proguard-rules.pro:9]: Hehe... somebody did a rename of name to courseName :)) It often happens that AS will replace it everywhere it finds it... :D

File: app/src/main/java/android/example/courses/CourseDetailFragment.kt (hunk 2 above)
Comment (radumarin88) [reply app/src/main/java/android/example/courses/CourseDetailFragment.kt:28]: It should be fairly straightforward: add a framelayout with the content being your first child and the progress indicator being its second child. Use View.VISIBLE and View.GONE (not INVISIBLE - invisible means you can't see it, but it still occupies space on the screen).

File: app/src/main/java/android/example/courses/MainActivity.kt [hunk 6]
```diff
@@ -4,16 +4,49 @@ import android.content.Intent
...
+        if (email == "" || pass == "") {
+            invalidCredentials()
+            return
+        }
+
+        auth.signInWithEmailAndPassword(email, pass)
+            .addOnCompleteListener(this) { task ->
+                if (task.isSuccessful) {
+                    val intent = Intent(this, NavigationActivity::class.java)
+                    startActivity(intent)
```
Comment (radumarin88) [inline app/src/main/java/android/example/courses/MainActivity.kt:35]: You are not calling finish() here which means that when the user presses back from the NavigationActivity, she will be back on the Login screen. Is this desired behavior?

File: app/src/main/java/android/example/courses/NavigationActivity.kt (hunk 3 above)
Comment (radumarin88) [reply app/src/main/java/android/example/courses/NavigationActivity.kt:77]: Picasso has something called placeholders. More about it on the doc page or even more here: https://futurestud.io/tutorials/picasso-placeholders-errors-and-fading

File: app/src/main/java/android/example/courses/AttendanceFragment.kt (hunk 1 above)
Comment (radumarin88) [reply app/src/main/java/android/example/courses/AttendanceFragment.kt]: Welcome to the world of UI-driven data! :D Firebase is an awesome tool that lets you have client-driven functionality, but at the same time you discover you need to plan it out nicely from the beginning. The path you took is good.
Now here is where I hate Google for adding things like FirestoreRecyclerAdapter... because it tightly couples your networking code with your business logic with the UI logic. What I would recommend is:
1. You move to using MVVM as taught in the last lab
2. Have the ViewModel do the queries and expose them as LiveData. See this for a beginner's reference to hiding Firebase behind LiveData: https://firebase.googleblog.com/2017/12/using-android-architecture-components.html
3. In the ViewModel when you get the results from the query you can actually map them to a data structure that's suited for being displayed in the UI - i.e. a list of objects that can be mapped directly to ViewHolders
4. In the Fragment, ditch the FirestoreRecyclerAdapter and just create your own RecylerView.Adapter that knows how to display the list from point 3. You won't need to worry about Firebase in the UI, you just observer a LiveData that exposes a List of data models and just pass that list to the adapter.

And this is your crash course in clean architecture :)

Btw, congrats on the project! I like what I see and you already have a grade of 10. If you move to MVVM, you'll also be getting a bonus.

Comment (cosmin-ionita) [reply app/src/main/java/android/example/courses/AttendanceFragment.kt]: @radumarin88: Thanks a lot for reviewing this PR and also for the kind words! And also for the grade!!

I gave a shot in the last commit of this PR to the MVVM approach by following the link you provided and it really changes the things significantly. LiveData decouples the data layer from the UI layer in an even-driven, lifecycle-aware way, which is pretty cool!

Because of time constraints, I only adapted to MVVM 2 (simpler) fragments: GradesFragment and AttendanceFragment, but I managed to get the overall idea.

This opportunity also empowered me to read about Android Architecture Components, which was something I really searched for a while (I wanted to see some best practices in Android development), so thanks a lot for this! It really helped me!

File: app/build.gradle (hunk 4 above)
Comment (cosmin-ionita) [reply app/build.gradle:61]: I added this file now in the last commit. I searched on the web about the safety of that file and I have some mixed feelings about it. Some say it's safe, others say it's not. I need to do some more research to see exactly what an attacker could do with this file.

In the meantime, I made this repo private, just in case. My firestore rules allow all access to the db, so it's better to be safe.

File: app/src/main/java/android/example/courses/MainActivity.kt (hunk 6 above)
Comment (cosmin-ionita) [reply app/src/main/java/android/example/courses/MainActivity.kt:35]: Great point, I missed this call. Added in the last commit. Thanks! 👍

**Comment by cosmin-ionita** [comment]
@radumarin88: I think you can test the app now. Please make sure to use this branch for testing.

The credentials you can use are: email: `ionitacosmincpp@gmail.com` and pass: `testtest`.

Also, if you want to add you on firestore to be able to see the new db structure, please give me your e-mail address.
